File: AWS/iac-microbenchmark/ec2/python/aes256_encrypt.py
#!/usr/bin/env python3
import boto3
import base64
import os
import secrets
import sys
import json
# Initialize encryption using AES-GCM
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

def main():
    # Initialize KMS client
    kms_client = boto3.client('kms')

    # Get the KMS key ARN from environment variables
    kms_key_id = os.environ['AES_KMS_KEY_ARN']

    # Read JSON input from stdin
    input_data = sys.stdin.read()

    # Parse the JSON input
    data = json.loads(input_data)

    # Extract the message from the event payload
    message = data.get('message')

    # Create a new data key for AES-256-GCM encryption
    response = kms_client.generate_data_key(
        KeyId=kms_key_id,
        KeySpec='AES_256'
    )
    
    plaintext_data_key = response['Plaintext']
    encrypted_data_key = response['CiphertextBlob']

    # Generate 16 secure random bytes
    iv = secrets.token_bytes(16)  # 16 bytes for AES-GCM IV

    cipher = Cipher(algorithms.AES(plaintext_data_key), modes.GCM(iv))
    encryptor = cipher.encryptor()

    # Encrypt the message (no padding required in AES-GCM)
    ciphertext = encryptor.update(message.encode('utf-8')) + encryptor.finalize()
    tag = encryptor.tag
    
    # Return the response without json.dumps
    encrypt_result =  {
        'encrypted_data_key': base64.b64encode(encrypted_data_key).decode('utf-8'),
        'iv': base64.b64encode(iv).decode('utf-8'),
        'tag': base64.b64encode(tag).decode('utf-8'),
        'encrypted_message': base64.b64encode(ciphertext).decode('utf-8')
    }

    print(json.dumps(encrypt_result))

if __name__ == "__main__":
    main()

File: AWS/iac-microbenchmark/ec2/typescript/aes256_encrypt.ts
#!/usr/bin/env node

import { KMSClient, GenerateDataKeyCommand } from '@aws-sdk/client-kms';
import * as process from 'process';
import * as crypto from 'crypto';
import * as base64 from 'base64-js';

// Initialize the AWS KMS client with the specified region
const kmsClient = new KMSClient({
  region: 'us-east-1' // Specify the desired region
});

async function main(): Promise<void> {
  try {
    // Get the KMS key ARN from environment variables
    const kmsKeyId = process.env['AES_KMS_KEY_ARN'];

    // Check if the KMS key ARN is provided
    if (!kmsKeyId) {
      throw new Error("AES_KMS_KEY_ARN environment variable is not set");
    }

    // Get the JSON input from command line arguments
    const requestJsonRaw: string = process.argv[2];

    // Parse the JSON input
    const requestJson = JSON.parse(requestJsonRaw);

    // Extract the message from the parsed JSON
    const message: string = requestJson["message"];

    // Create a new data key for AES-256-GCM encryption
    const generateDataKeyCommand = new GenerateDataKeyCommand({
      KeyId: kmsKeyId,
      KeySpec: 'AES_256',
    });

    const response = await kmsClient.send(generateDataKeyCommand);

    const plaintextDataKey = response.Plaintext as Uint8Array; // The plaintext data key
    const encryptedDataKey = response.CiphertextBlob as Uint8Array; // The encrypted data key

    const iv = crypto.randomBytes(12); // 12 bytes for AES-GCM IV

    // Create a cipher using AES-GCM
    const cipher = crypto.createCipheriv('aes-256-gcm', plaintextDataKey, iv);

    // Encrypt the message (no padding required in AES-GCM)
    const ciphertext = Buffer.concat([
      cipher.update(Buffer.from(message, 'utf-8')),
      cipher.final()
    ]);

    // Get the authentication tag
    const tag = cipher.getAuthTag();

    // Prepare the result
    const encryptResult = {
      encrypted_data_key: base64.fromByteArray(encryptedDataKey),
      iv: base64.fromByteArray(iv),
      tag: base64.fromByteArray(tag),
      encrypted_message: base64.fromByteArray(ciphertext)
    };

    // Output the result as a JSON string
    console.log(JSON.stringify(encryptResult));
  } catch (error) {
    // Handle and log any errors that occur
    console.error("Error:", error);
  }
}

// Execute the main function
main();
